src/celeritas/neutron/model/NeutronInelasticModel.cc

```diff
@@ -0,0 +1,144 @@
+
+    CELER_VALIDATE(data.ids.neutron,
+                   << "missing neutron particles (required for "
+                   << this->description() << ")");
+
+    // Save particle properties
+    data.neutron_mass = particles.get(data.ids.neutron).mass();
+
+    // Load neutron elastic cross section data
+    make_builder(&data.micro_xs).reserve(materials.num_elements());
+    for (auto el_id : range(ElementId{materials.num_elements()}))
+    {
+        AtomicNumber z = materials.get(el_id).atomic_number();
+        this->append_xs(load_data(z), &data);
+    }
+    CELER_ASSERT(data.micro_xs.size() == materials.num_elements());
+
+    // Move to mirrored data, copying to device
+    mirror_ = CollectionMirror<NeutronInelasticData>{std::move(data)};
+    CELER_ENSURE(this->mirror_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Particle types and energy ranges that this model applies to.
+ */
+auto NeutronInelasticModel::applicability() const -> SetApplicability
+{
+    Applicability neutron_applic;
+    neutron_applic.particle = this->host_ref().ids.neutron;
+    neutron_applic.lower = this->host_ref().min_valid_energy();
+    neutron_applic.upper = this->host_ref().max_valid_energy();
+
+    return {neutron_applic};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the microscopic cross sections for the given particle and material.
+ */
+auto NeutronInelasticModel::micro_xs(Applicability) const -> MicroXsBuilders
+{
+    // Cross sections are calculated on the fly
+    return {};
+}
+
+//---------------------------------------------------------------------------//
+//!@{
+/*!
+ * Apply the interaction kernel.
+ */
+void NeutronInelasticModel::execute(CoreParams const& params,
+                                    CoreStateHost& state) const
+{
+    auto execute = make_action_track_executor(
+        params.ptr<MemSpace::native>(),
+        state.ptr(),
+        this->action_id(),
+        InteractionApplier{NeutronInelasticExecutor{this->host_ref()}});
+    return launch_action(*this, params, state, execute);
+}
+
+//---------------------------------------------------------------------------//
+#if !CELER_USE_DEVICE
+void NeutronInelasticModel::execute(CoreParams const&, CoreStateDevice&) const
+{
+    CELER_NOT_CONFIGURED("CUDA OR HIP");
+}
+#endif
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the model ID for this model.
+ */
+ActionId NeutronInelasticModel::action_id() const
+{
+    return this->host_ref().ids.action;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct neutron inelastic cross section data for a single element.
+ */
+void NeutronInelasticModel::append_xs(ImportPhysicsVector const& inp,
+                                      HostXsData* data) const
+{
+    auto reals = make_builder(&data->reals);
```

Comment:
We may want to refactor the construction so that the grids can be deduplicated (if multiple materials use the same energy grid, for example).  See `orange/detail/UnitInserter.hh` and other classes in that directory as an example. We could trivially add a class to be used here (and in other params classes that build grid data):
```c++
class GridInserter
{
  public:
    //!@{
    //! \name Type aliases
    template<class T>
    using Items = Collection<T, Ownership::value, MemSpace::host>;
    using GridId = ItemId<GenericGridData>;
    //!@}

  public:
    GridInserter(Items<GenericGridData>* grids, Items<real_type>* reals);

    GridId operator()(ImportPhysicsVector const&);

  private:
    CollectionBuilder<GenericGridData> grids_;
    DedupeCollectionBuilder<real_type> reals_;
};
```
but one of us could do this in a follow-on PR.

Reply:
In fact I'm working on this right now.

Reply:
Thanks for another PR, [#1157](https://github.com/celeritas-project/celeritas/pull/1157)  on this!